lnwallet/size.go

```diff
@@ -301,3 +321,97 @@ func estimateCommitTxWeight(count int, prediction bool) int64 {
 
 	return htlcWeight + baseWeight + witnessWeight
 }
+
+// TxWeightEstimator is able to calculate weight estimates for transactions
+// based on the input and output types. For purposes of estimation, all
+// signatures are assumed to be of the maximum possible size, 73 bytes.
+type TxWeightEstimator struct {
+	hasWitness       bool
+	inputCount       uint32
+	outputCount      uint32
+	inputSize        int
+	inputWitnessSize int
+	outputSize       int
+}
+
+// AddP2PKHInput updates the weight estimate to account for an additional input
+// spending a P2PKH output.
+func (twe *TxWeightEstimator) AddP2PKHInput() {
+	twe.inputSize += InputSize + P2PKHScriptSigSize
+	twe.inputWitnessSize++
+	twe.inputCount++
+}
+
+// AddP2WKHInput updates the weight estimate to account for an additional input
+// spending a native P2PWKH output.
+func (twe *TxWeightEstimator) AddP2WKHInput() {
+	twe.AddWitnessInput(P2WKHWitnessSize)
+}
+
+// AddWitnessInput updates the weight estimate to account for an additional
+// input spending a native pay-to-witness output. This accepts the total size
+// of the witness as a parameter.
+func (twe *TxWeightEstimator) AddWitnessInput(witnessSize int) {
+	twe.inputSize += InputSize
+	twe.inputWitnessSize += witnessSize
+	twe.inputCount++
+	twe.hasWitness = true
+}
+
+// AddNestedP2WKHInput updates the weight estimate to account for an additional
+// input spending a P2SH output with a nested P2WKH redeem script.
+func (twe *TxWeightEstimator) AddNestedP2WKHInput() {
+	twe.inputSize += InputSize + P2WPKHSize
+	twe.inputWitnessSize += P2WKHWitnessSize
+	twe.inputSize++
+	twe.hasWitness = true
+}
+
+// AddNestedP2WSHInput updates the weight estimate to account for an additional
+// input spending a P2SH output with a nested P2WSH redeem script.
+func (twe *TxWeightEstimator) AddNestedP2WSHInput(witnessSize int) {
+	twe.inputSize += InputSize + P2WSHSize
+	twe.inputWitnessSize += witnessSize
+	twe.inputSize++
+	twe.hasWitness = true
+}
+
+// AddP2PKHOutput updates the weight estimate to account for an additional P2PKH
+// output.
+func (twe *TxWeightEstimator) AddP2PKHOutput() {
+	twe.outputSize += P2PKHOutputSize
+	twe.outputCount++
+}
+
+// AddP2WKHOutput updates the weight estimate to account for an additional
+// native P2WKH output.
+func (twe *TxWeightEstimator) AddP2WKHOutput() {
+	twe.outputSize += P2WKHOutputSize
+	twe.outputCount++
+}
+
+// AddP2WSHOutput updates the weight estimate to account for an additional
+// native P2WSH output.
+func (twe *TxWeightEstimator) AddP2WSHOutput() {
+	twe.outputSize += P2WSHOutputSize
+	twe.outputCount++
+}
+
+// AddP2SHOutput updates the weight estimate to account for an additional P2SH
+// output.
+func (twe *TxWeightEstimator) AddP2SHOutput() {
+	twe.outputSize += P2SHOutputSize
+	twe.outputCount++
+}
+
+// Weight gets the estimated weight of the transaction.
+func (twe *TxWeightEstimator) Weight() int {
```

Comment:
The output type here is `int` to be consistent with the input types of the witness weights in `AddWitnessInput` & `AddNestedP2WSHInput`. Basically, I think it's best if there is one weight unit. Unfortunately btcd is not all that consistent, but `int` was slightly easier to use because all of the constants at the top are implicitly declared as `int` and `wire.VarIntSerializeSize` returns an int.

Generally in the places where weight gets cast to `uint64`, it is in the context of converting from weight to `btcutil.Amount` and it seems like an OK thing to cast there since weight and satoshis are different logical units.

Reply:
Ahh yes, 100% agree with you. Would be nice to standardize a weight type, but seems beyond the scope of the PR given the other constraints, e.g. constants defined as `int`s, func return types, etc. Can just be something we keep in mind as potential future improvements!